staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook_test.go

```diff
@@ -0,0 +1,321 @@
+
+	s := httptest.NewServer(newWebhookHandler(t, func(events *auditv1alpha1.EventList) {
+		gotEvents = true
+	}))
+	defer s.Close()
+
+	backend := newTestBlockingWebhook(t, s.URL)
+
+	// Ensure this doesn't return a serialization error.
+	event := &auditinternal.Event{}
+	require.NoError(t, backend.processEvents(event), "failed to send events")
+}
+
+// waitForEmptyBuffer indicates when the sendBatchEvents method has read from the
+// existing buffer. This lets test coordinate closing a timer and stop channel
+// until the for loop has read from the buffer.
+func waitForEmptyBuffer(b *batchBackend) {
+	for len(b.buffer) != 0 {
+		time.Sleep(time.Millisecond)
+	}
+}
+
+func TestBatchWebhookMaxEvents(t *testing.T) {
+	nRest := 10
+	events := make([]*auditinternal.Event, defaultBatchMaxSize+nRest) // greater than max size.
+	for i := range events {
+		events[i] = &auditinternal.Event{}
+	}
+
+	got := make(chan int, 2)
+	s := httptest.NewServer(newWebhookHandler(t, func(events *auditv1alpha1.EventList) {
+		got <- len(events.Items)
+	}))
+	defer s.Close()
+
+	backend := newTestBatchWebhook(t, s.URL)
+
+	backend.ProcessEvents(events...)
+
+	stopCh := make(chan struct{})
+	timer := make(chan time.Time, 1)
+
+	backend.sendBatchEvents(stopCh, timer)
+	require.Equal(t, defaultBatchMaxSize, <-got, "did not get batch max size")
+
+	go func() {
+		waitForEmptyBuffer(backend) // wait for the buffer to empty
+		timer <- time.Now()         // Trigger the wait timeout
+	}()
+
+	backend.sendBatchEvents(stopCh, timer)
+	require.Equal(t, nRest, <-got, "failed to get the rest of the events")
+}
+
+func TestBatchWebhookStopCh(t *testing.T) {
+	events := make([]*auditinternal.Event, 1) // less than max size.
+	for i := range events {
+		events[i] = &auditinternal.Event{}
+	}
+
+	expected := len(events)
+	got := make(chan int, 2)
+	s := httptest.NewServer(newWebhookHandler(t, func(events *auditv1alpha1.EventList) {
+		got <- len(events.Items)
+	}))
+	defer s.Close()
+
+	backend := newTestBatchWebhook(t, s.URL)
+	backend.ProcessEvents(events...)
+
+	stopCh := make(chan struct{})
+	timer := make(chan time.Time)
+
+	go func() {
+		waitForEmptyBuffer(backend)
+		close(stopCh) // stop channel has stopped
+	}()
+	backend.sendBatchEvents(stopCh, timer)
+	require.Equal(t, expected, <-got, "get queued events after timer expires")
+}
+
+func TestBatchWebhookEmptyBuffer(t *testing.T) {
+	got := make(chan bool, 1)
+	s := httptest.NewServer(newWebhookHandler(t, func(events *auditv1alpha1.EventList) {
+		got <- true
+	}))
+	defer s.Close()
+
+	backend := newTestBatchWebhook(t, s.URL)
+
+	stopCh := make(chan struct{})
+	timer := make(chan time.Time)
+
+	close(stopCh) // stop channel has stopped
+
+	// Buffer is empty, no events have been queued. This should exit but send no events.
+	backend.sendBatchEvents(stopCh, timer)
+
+	go func() {
+		// Wait a bit to ensure the webhook did not send.
+		time.Sleep(500 * time.Millisecond)
```

Comment:
Good idea, we're really testing that the webhook didn't send a POST with zero events anyway. Updated.